sklearn/utils/validation.py

```diff
@@ -1120,6 +1127,12 @@ def is_sparse(dtype):
             else:
                 array = array.copy(**copy_params)
 
+    if ensure_non_negative:
+        whom = input_name
+        if estimator_name:
+            whom += f" in {estimator_name}"
+        check_non_negative(array, whom)
+
```

Comment:
I would put that just before the previous ``if force_writeable`` block to be able to fail fast, before a pontential copy.